### PR TITLE
DCOS-19648: add placement constraint validator

### PR DESCRIPTION
**Pull request by bstavroulakis**
1) Add placement constraint validator to make sure the operator/ field combo is unique.

image

2) These errors are "permissive" meaning that you can post the form without them being satisfied. And they are hidden from the ErrorsAlert by default.

Closes DCOS-19648

**Checklist**
- [ ] Did you add a JIRA issue in a commit message or as part of the branch name?
- [ ] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

**Comment by juliangieseke** [comment]
this needs some context, why is help needed?

File: plugins/services/src/js/validators/PlacementsValidators.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ function checkDuplicateOperatorField(constraints) {
 
     if (visitedOperatorFieldPairs.indexOf(key) !== -1) {
       errors.push(
-        [].concat([
```
Comment (DanielMSchmidt) [inline plugins/services/src/js/validators/PlacementsValidators.js]: We could als do `return [...errors, {...}, {...}]`, I think that would be a bit easier to read 🤔

File: src/js/components/JSONEditor.js [hunk 2]
```diff
@@ -407,7 +407,9 @@ class JSONEditor extends React.Component {
       ];
     }
 
-    return this.externalErrors.map(error => {
+    const errors = [].concat.apply([], this.externalErrors);
```
Comment (ahoskins) [inline src/js/components/JSONEditor.js]: Why this change? It's already doing `(this.props.errors || []).slice();` whenever it's set.

File: src/js/components/ErrorsAlert.js [hunk 3]
```diff
@@ -4,8 +4,17 @@ import Alert from "./Alert";
...
 const ErrorsAlert = function(props) {
-  const { errors, hideTopLevelErrors, pathMapping } = props;
-  let showErrors = errors;
+  const {
+    errors,
+    hideTopLevelErrors,
+    hidePermissiveErrors,
+    pathMapping
+  } = props;
+  let showErrors = [].concat.apply([], errors);
```
Comment (ahoskins) [inline src/js/components/ErrorsAlert.js]: Why is this needed. Default prop type is `[]` so this seems unneeded.

File: plugins/services/src/js/validators/PlacementsValidators.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      ? { fieldName: constraint.fieldName, operator: constraint.operator }
+      : { fieldName: constraint[0], operator: constraint[1] };
+    const key = `{${operatorFieldPair.operator}}{${operatorFieldPair.fieldName}}`;
+
+    if (visitedOperatorFieldPairs.indexOf(key) !== -1) {
+      errors.push(
+        ...[
+          {
+            path: ["constraints", index, "fieldName"],
+            message: "Duplicate operator/ field set",
```
Comment (ahoskins) [inline plugins/services/src/js/validators/PlacementsValidators.js]: Should add a space before the /

File: plugins/services/src/js/validators/PlacementsValidators.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            isPermissive: true
+          },
+          {
+            path: ["constraints", index, "operator"],
+            message: "Duplicate operator/ field set",
```
Comment (ahoskins) [inline plugins/services/src/js/validators/PlacementsValidators.js]: And here

File: plugins/services/src/js/components/modals/CreateServiceModal.js [hunk 6]
```diff
@@ -456,7 +462,10 @@ class CreateServiceModal extends Component {
 
   handleServiceReview() {
     const errors = this.getFormErrors();
-    if (errors.length === 0) {
+    if (
```
Comment (ahoskins) [inline plugins/services/src/js/components/modals/CreateServiceModal.js]: To make this more clear, maybe condense into a single expression reading "are there any not permissive errors?":

```
if (errors.filter(error => !error.isPermissive).length > 0) {}
```

**Comment by weblancaster** [comment]
@bstavroulakis is `DCOS-19649` the correct Jira ticket? I checked the Jira for more description and there is nothing related to this PR

File: plugins/services/src/js/validators/PlacementsValidators.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import ValidatorUtil from "#SRC/js/utils/ValidatorUtil";
+import { findNestedPropertyInObject } from "#SRC/js/utils/Util";
+import PlacementConstraintsUtil from "../utils/PlacementConstraintsUtil";
+import { PROP_MISSING_ONE, SYNTAX_ERROR } from "../constants/ServiceErrorTypes";
+
+function checkDuplicateOperatorField(constraints) {
+  if (!constraints || ValidatorUtil.isEmpty(constraints)) {
```
Comment (weblancaster) [inline plugins/services/src/js/validators/PlacementsValidators.js]: No need for `!constraints` because `ValidatorUtil.isEmpty` already checks for `null` and undefined.

File: plugins/services/src/js/validators/PlacementsValidators.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+  const errors = [];
+  const visitedOperatorFieldPairs = [];
+  constraints.map(function(constraint, index) {
+    const operatorFieldPair = "operator" in constraint
+      ? { fieldName: constraint.fieldName, operator: constraint.operator }
+      : { fieldName: constraint[0], operator: constraint[1] };
+    const key = `{${operatorFieldPair.operator}}{${operatorFieldPair.fieldName}}`;
+
+    if (visitedOperatorFieldPairs.indexOf(key) !== -1) {
```
Comment (weblancaster) [inline plugins/services/src/js/validators/PlacementsValidators.js]: For readability `visitedOperatorFieldPairs .includes(key)`

File: src/js/components/ErrorsAlert.js (hunk 3 above)
Comment (weblancaster) [reply src/js/components/ErrorsAlert.js]: @ahoskins are you saying the default `errors` prop is `[]`?

Comment (weblancaster) [reply src/js/components/ErrorsAlert.js]: I do see the case to keep `[].concat.apply([], errors)` in case `errors` is `undefined` or `null` which won't throw error when `showErrors.map..` or alike calls

File: src/js/components/JSONEditor.js (hunk 2 above)
Comment (weblancaster) [reply src/js/components/JSONEditor.js]: @ahoskins please see my comment above https://github.com/dcos/dcos-ui/pull/2539/files#r161327729

File: src/js/components/ErrorsAlert.js (hunk 3 above)
Comment (ahoskins) [reply src/js/components/ErrorsAlert.js]: Okay, fair enough. When it's `undefined` no issue because it'll use the default prop. But to protect against `null` this is needed. 👍

**Comment by juliangieseke** [comment]
Alfred test this

**Comment by juliangieseke** [comment]
'flicts

**Comment by juliangieseke** [comment]
Alfred test this

**Comment by juliangieseke** [comment]
Alfred test this

**Comment by bstavroulakis** [comment]
@weblancaster Sorry changed to DCOS-19648

**Comment by bstavroulakis** [comment]
@ahoskins 
1) You are referring to a separate error concerning duplicate values. If you select another field with a different value the permissive errors will kick in.
2) Fixed for frameworks as well.

**Comment by bstavroulakis** [comment]
Alfred test this

**Comment by juliangieseke** [comment]
Alfred test this

**Comment by DanielMSchmidt** [comment]
Alfred test this please :)

**Comment by bstavroulakis** [comment]
Alfred my friend, please test this

**Comment by bstavroulakis** [comment]
Alfred test this

**Comment by Poltergeist** [comment]
Alfred you can't test this. 👯 👯‍♂️ 🎉

**Comment by DanielMSchmidt** [comment]
I found two cases that I (with my limited knowledge of the system and constraints) think might be not ideal:

1) showing two errors to the user
bildschirmfoto 2018-01-17 um 15 08 12

2) validating on blur
placement-validation-after-blur

File: plugins/services/src/js/components/modals/CreateServiceModalForm.js [hunk 9]
```diff
@@ -647,8 +647,9 @@ class CreateServiceModalForm extends Component {
   getUnmutedErrors() {
     const { showAllErrors } = this.props;
     const { editedFieldPaths, editingFieldPath } = this.state;
+    const errors = [].concat.apply([], this.getErrors());
```
Comment (DanielMSchmidt) [inline plugins/services/src/js/components/modals/CreateServiceModalForm.js]: I would assume from the name `getErrors` returns an array, why do we need this super complex looking `[].concat.apply([],` stuff around it?

Comment (DanielMSchmidt) [reply plugins/services/src/js/components/modals/CreateServiceModalForm.js]: Found [this](https://github.com/dcos/dcos-ui/pull/2539/files#r161301091), makes sense

**Comment by bstavroulakis** [comment]
Alfred test this

File: src/styles/components/form-elements/form-components/select/styles.less [hunk 10]
```diff
@@ -1,5 +1,11 @@
 & when (@select-enabled) {
   /*  Select (.dropdown-select) */
+  .form-group.form-group-danger {
+
+    .dropdown-toggle {
+      border-color: @red !important;
```
Comment (weblancaster) [inline src/styles/components/form-elements/form-components/select/styles.less]: Discussed offline. Removing the `important`

**Comment by weblancaster** [comment]
alfred test this

**Comment by bstavroulakis** [comment]
Alfred please please please test this

**Comment by bstavroulakis** [comment]
Alfred test this please please please

**Comment by bstavroulakis** [comment]
Alfred test this 🥇

**Comment by DanielMSchmidt** [comment]
Alfred test this please :)

**Comment by Poltergeist** [comment]
Alfred you can't test this. 👯 👯‍♂️ 🎉

**Comment by nLight** [comment]
Alfred test this

**Comment by Poltergeist** [comment]
Alfred you can't test this. 👯 👯‍♂️ 🎉

**Comment by juliangieseke** [comment]
Alfred test this
last: https://jenkins.mesosphere.com/service/jenkins/job/Frontend/job/dcos-ui-pull-requests/2314/

**Comment by DanielMSchmidt** [comment]
Alfred test this

Failed because of "No test report files were found" after jest
